### PR TITLE
Adding autocomplete

### PR DESCRIPTION
**Pull request by ronykoz**
Changing the SDK to work with click library. adding the option to accomplish autocomplete in the CLI.

To enable the autocomplete run in the terminal `eval "$(_DEMISTO_SDK_COMPLETE=source demisto_sdk)"`
or if using zsh then `eval "$(_DEMISTO_SDK_COMPLETE=source_zsh demisto-sdk)"`

### CODE REVIEW

**Comment by ronykoz** [comment]
@roysagi FYI

**Comment by ronykoz** [comment]
Once this PR is merge we will need to make some change to the build in Content repo too

File: demisto_sdk/main.py [hunk 1]
```diff
@@ -1,11 +1,241 @@
...
+@click.option(
+    '-c', '--circle', type=click.Choice(["True", "False"]), default='False',
+    help='Is CircleCi or not')
+@click.option(
+    '-wl', '--whitelist', default='./Tests/secrets_white_list.json',
+    help='Full path to whitelist file, file name should be "secrets_white_list.json"')
+@pass_config
+def secrets(config, **kwargs):
+    sys.path.append(config.configuration.env_dir)
+    validator = SecretsValidator(configuration=config.configuration, is_circle=str2bool(kwargs['circle']),
```
Comment (yaakovi) [inline demisto_sdk/main.py]: why `secret_white_list.json` has a configuration but `conf.json` and `id_set.json` in validate doesn't?

File: demisto_sdk/main.py [hunk 2]
```diff
@@ -1,11 +1,241 @@
...
+    '-h', '--help'
+)
+@click.option(
+    '-a', '--artifacts_path', help='The path of the directory in which you want to save the created content artifacts')
+@click.option(
+    '-p', '--preserve_bundles', is_flag=True, default=False,
+    help='Flag for if you\'d like to keep the bundles created in the process of making the content artifacts')
+def create(**kwargs):
+    content_creator = ContentCreator(**kwargs)
+    content_creator.run()
```
Comment (yaakovi) [inline demisto_sdk/main.py]: should this return a value? (same as rest of the commands)

File: demisto_sdk/main.py [hunk 3]
```diff
@@ -1,11 +1,241 @@
...
+    linter = Linter(configuration=config.configuration, project_dir=dir, **kwargs)
+    return linter.run_dev_packages()
 
 
-def main():
-    # TODO: Typings and docstrings
-    sdk = DemistoSDK()
-    return sdk.parse_args()
+def demisto_sdk_cli():
+    print(main())
```
Comment (yaakovi) [inline demisto_sdk/main.py]: verify that command failure causing error code to be 1 (don't revive previous errors)

File: demisto_sdk/validation/file_validator.py [hunk 4]
```diff
@@ -522,18 +531,8 @@ def validate_against_previous_version(self, no_error=False):
...
-        parser.add_argument('-c', '--circle', type=str2bool, default=False, help='Is CircleCi or not')
-        parser.add_argument('-b', '--backward-comp', type=str2bool, default=True,
-                            help='To check backward compatibility.')
-        parser.add_argument('-t', '--test-filter', type=str2bool, default=False,
-                            help='Check that tests are valid.')
-        parser.add_argument('-j', '--conf-json', action='store_true', help='Validate the conf.json file.')
-        parser.add_argument('-i', '--id-set', action='store_true', help='Create the id_set.json file.')
-        parser.add_argument('-p', '--prev-ver', help='Previous branch or SHA1 commit to run checks against.')
-        parser.add_argument('-g', '--use-git', action='store_true', help='Validate changes using git.')
+    # parser.add_argument('-t', '--test-filter', type=str2bool, default=False,
```
Comment (yaakovi) [inline demisto_sdk/validation/file_validator.py]: TODO.

after validation there was a step to run the `configure_tests` script to check that each changed file had a relevant test - was used as part of the hooks.

File: demisto_sdk/yaml_tools/content_creator.py [hunk 5]
```diff
@@ -57,6 +60,12 @@ def __init__(self, artifacts_path: str, content_bundle_path='',
         self.file_name_max_size = 85
         self.long_file_names = []  # type:List
 
+    def run(self):
```
Comment (yaakovi) [inline demisto_sdk/yaml_tools/content_creator.py]: should have indication of failures (return)

File: demisto_sdk/yaml_tools/extractor.py [hunk 6]
```diff
@@ -168,22 +173,22 @@ def get_yml_type(self) -> str:
...
-        parser.add_argument("-t", "--type",
-                            help="Yaml type. If not specified will try to determine type based upon path.",
-                            choices=[SCRIPT_PREFIX, INTEGRATION_PREFIX], default=None)
-        parser.add_argument("-d", "--demistomock", help="Add an import for demisto mock, true by default",
-                            choices=[True, False], type=str2bool, default=True)
-        parser.add_argument("-c", "--commonserver",
-                            help=("Add an import for CommonServerPython. "
-                                  " If not specified will import unless this is CommonServerPython"),
-                            choices=[True, False], type=str2bool, default=None)
+    # @staticmethod
```
Comment (yaakovi) [inline demisto_sdk/yaml_tools/extractor.py]: remove

File: demisto_sdk/yaml_tools/unifier.py [hunk 7]
```diff
@@ -6,22 +6,32 @@
...
 
 
 class Unifier:
 
-    def __init__(self, package_path: str, dir_name=INTEGRATIONS_DIR, dest_path='', image_prefix=DEFAULT_IMAGE_PREFIX):
+    def __init__(self, indir: str, dir_name=INTEGRATIONS_DIR, outdir='',
+                 image_prefix=DEFAULT_IMAGE_PREFIX):
+
+        directory_name = ""
+        for optional_dir_name in DIR_TO_PREFIX.keys():
```
Comment (yaakovi) [inline demisto_sdk/yaml_tools/unifier.py]:
```suggestion
        for optional_dir_name in DIR_TO_PREFIX:
```

File: demisto_sdk/yaml_tools/unifier.py [hunk 8]
```diff
@@ -71,9 +81,10 @@ def write_yaml_with_docker(self, yml_text, yml_data, script_obj):
             }
         for file_path, file_text in output_map.items():
             if os.path.isfile(file_path):
-                raise ValueError('Output file already exists: {}.'
-                                 ' Make sure to remove this file from source control'
-                                 ' or rename this package (for example if it is a v2).'.format(self.dest_path))
+                # raise ValueError('Output file already exists: {}.'
```
Comment (yaakovi) [inline demisto_sdk/yaml_tools/unifier.py]: ?

File: demisto_sdk/main.py [hunk 9]
```diff
@@ -1,11 +1,241 @@
...
+@click.option(
+    "--no-bandit", is_flag=True, help="Do NOT run bandit linter")
+@click.option(
+    "--no-test", is_flag=True, help="Do NOT test (skip pytest)")
+@click.option(
+    "-r", "--root", is_flag=True, help="Run pytest container with root user")
+@click.option(
+    "-k", "--keep-container", is_flag=True, help="Keep the test container")
+@click.option(
+    "-v", "--verbose", is_flag=True, help="Verbose output")
```
Comment (anara123) [inline demisto_sdk/main.py]: in the help we should mention that this flag is for debugging/troubleshooting purposes

File: demisto_sdk/main.py [hunk 10]
```diff
@@ -1,11 +1,241 @@
...
+    '--common-server', '-c',
+    help="Add an import for CommonServerPython."
+         "If not specified will import unless this is CommonServerPython",
+    type=click.Choice(["True", "False"]),
+    default='False'
+)
+@pass_config
+def extract(config, **kwargs):
+    extractor = Extractor(configuration=config.configuration, **kwargs)
+    return extractor.extract_to_package_format()
```
Comment (anara123) [inline demisto_sdk/main.py]: it should also be `run`

File: demisto_sdk/yaml_tools/extractor.py [hunk 11]
```diff
@@ -110,9 +113,8 @@ def migrate(self) -> int:
               )
         return 0
 
-    def extract_code(self, output_file) -> int:
+    def extract_code(self, code_file_path) -> int:
         yml_type = self.get_yml_type()
-        print("Extracting code to: {} ...".format(self.dest_path))
         common_server = self.common_server
         if common_server is None:
```
Comment (anara123) [inline demisto_sdk/yaml_tools/extractor.py]: common_server couldnt be None now. Either True or False.

File: demisto_sdk/main.py [hunk 12]
```diff
@@ -1,11 +1,241 @@
...
+    help="Add an import for demisto mock, true by default",
+    type=click.Choice(["True", "False"]),
+    default=True
+)
+@click.option(
+    '--common-server', '-c',
+    help="Add an import for CommonServerPython."
+         "If not specified will import unless this is CommonServerPython",
+    type=click.Choice(["True", "False"]),
+    default='False'
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/main.py]: Done

File: demisto_sdk/main.py [hunk 13]
```diff
@@ -1,11 +1,241 @@
...
+)
+@pass_config
+def extract(config, **kwargs):
+    extractor = Extractor(configuration=config.configuration, **kwargs)
+    return extractor.extract_to_package_format()
+
+
+@main.command(name="extract_code",
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/main.py]: felt that it is not user friendly

File: demisto_sdk/main.py (hunk 10 above)
Comment (ronykoz) [reply demisto_sdk/main.py]: does it really matter, i'm doing two different command with the same class which will make me pass another arg to run, i think that it's better this way

File: demisto_sdk/main.py (hunk 2 above)
Comment (ronykoz) [reply demisto_sdk/main.py]: not really

File: demisto_sdk/main.py (hunk 1 above)
Comment (ronykoz) [reply demisto_sdk/main.py]: should be an enhancement to those commands

File: demisto_sdk/yaml_tools/content_creator.py (hunk 5 above)
Comment (ronykoz) [reply demisto_sdk/yaml_tools/content_creator.py]: i think that it will just fail here

**Comment by ronykoz** [comment]
@yaakovi @anara123 can we merge this?

File: docs/development_guide.md [hunk 14]
```diff
@@ -26,6 +26,14 @@ You should call this method in the `initialize_parsers` method in the core SDK c
 All tests are run from the `tests` folder. They also run in the CircleCI build.
 Also make sure your methods work from the CLI by running `python demisto_sdk <your_method>` in your local environment.
 
+### How to run build using an unreleased deemisto-sdk version
```
Comment (yaakovi) [inline docs/development_guide.md]: demisto

**Pull request merged**